avr-hal-generic/src/port.rs

```diff
@@ -707,44 +709,46 @@ macro_rules! impl_port_traditional {
 
                     #[inline]
                     unsafe fn out_set(&mut self) {
-                        (*<$port>::ptr()).[<port $name:lower>].modify(|r, w| {
-                            w.bits(r.bits() | (1 << $pin))
+                        (*<$port>::ptr()).[<port $name:lower>].modify(|_, w| {
+                            w.[<p $name:lower $pin>]().set_bit()
                         })
                     }
 
                     #[inline]
                     unsafe fn out_clear(&mut self) {
-                        (*<$port>::ptr()).[<port $name:lower>].modify(|r, w| {
-                            w.bits(r.bits() & !(1 << $pin))
+                        (*<$port>::ptr()).[<port $name:lower>].modify(|_, w| {
+                            w.[<p $name:lower $pin>]().clear_bit()
                         })
                     }
 
                     #[inline]
                     unsafe fn out_toggle(&mut self) {
-                        (*<$port>::ptr()).[<pin $name:lower>].write(|w| w.bits(1 << $pin))
+                        (*<$port>::ptr()).[<pin $name:lower>].modify(|_, w| {
+                            w.[<p $name:lower $pin>]().set_bit()
+                        })
                     }
```

Comment:
I'd prefer staying with `.write()`.  `.modify()` only works here because of the `sbi`-optimization.  If you'd spell it out in its entirety (=> read PINx, bitor mask, write PINx), the behavior would be completely wrong:  It would also toggle any other pins which happen to read HIGH at this moment.

I'm not comfortable with relying on the optimization for correctness (as I had mentioned in https://github.com/Rahix/avr-hal/issues/284#issuecomment-1167648030).  So while it'd be nice to generate an `sbi` for `out_toggle()`, I think we should lean on the safe side and accept the `ldi`-`out` as a slightly less efficient, but always correct, solution...

Reply:
I get that, I thought I had changed back to `.write`, but seems not. I wonder if there's a way to cause a write to generate `sbi`, though. In avr-device, the register specs have constants to tell whether each bit gets modified by 0, 1 or both. They currently say any write modifies the register, perhaps if we can patch the generated svds to have this corrected, the compiler will optimize better. I'll do some experimenting later.